src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -223,10 +204,39 @@ private void EnqueueWork(bool highPriority, bool frozenPartialSemantics, Cancell
             {
                 // Normal request to either compute frozen partial tags, or compute normal tags in a tagger that does
                 // *not* support frozen partial tagging.
-                return await RecomputeTagsAsync(highPriority, frozenPartialSemantics, cancellationToken).ConfigureAwait(false);
+                await RecomputeTagsAsync(highPriority, frozenPartialSemantics, calledFromJtfRun: false, cancellationToken).ConfigureAwait(false);
+                return default;
             }
         }
 
+        /// <summary>
+        /// Spins, repeatedly calling into <paramref name="callback"/> with the current state of the tag trees.  When
+        /// the result of the callback can be saved without any intervening writes to <see
+        /// cref="_cachedTagTrees_mayChangeFromAnyThread"/> happening on another thread, then this helper returns. This
+        /// helper may also returns <see langword="null"/> in the case of cancellation.
+        /// </summary>
+        private async Task<(ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> oldTagTrees, ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> newTagTrees)>
+            CompareAndSwapTagTreesAsync(
+            Func<ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>>, ValueTask<ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>>>> callback,
+            CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var oldTagTrees = _cachedTagTrees_mayChangeFromAnyThread;
+
+                // Compute the new tag trees, based on what the current tag trees are.  Intentionally CA(true) here so
+                // we stay on the UI thread if we're in a JTF blocking call.
+                var newTagTrees = await callback(oldTagTrees).ConfigureAwait(true);
```

Comment:
>ConfigureAwait(true); [](http://example.com/codeflow?start=62&length=21)

Can you walk me through this a bit, going off the common path with ConfigureAwait always confuses me.

Specifically, the case I'm curious about is when we enter this method on the main thread and the callback switches to a different thread to do it's work. 

With ConfigureAwait(true), we get switched back to the main thread upon the callback return, right? Whereas, if ConfigureAwait(false), we stay on whatever thread the work completed on (or is always a threadpool thread)?

If that's so, wouldn't that be an extra switch to main thread, perhaps unnecessarilly in the case where the CompareExchange was successful? Or does the caller of this method expect to be on the main thread upon it's return?

Reply:
> If that's so, wouldn't that be an extra switch to main thread, perhaps unnecessarilly in the case where the CompareExchange was successful? Or does the caller of this method expect to be on the main thread upon it's return?

The caller can be in one of two modes.

1. Normal threadpool/bg tagging.  In this case, CA(true) is fine, as we come back to that context and continue i nthe bg.
2. blocking JTF call waiting for tags, for certain blocking editor APis (specifically, code-folding).  In this case, we want to return to he ui thread whenever we can *since that thread is already blocking on us*.  thsi allows that true thread to contribute progress to this work, instead of it just sitting there blocked, waiting on teh theradpool to finish work.